src/ec2/model.ts

```diff
@@ -41,21 +42,39 @@ export class Ec2ConnectionManager {
         return new DefaultIamClient(this.regionCode)
     }
 
-    protected async getAttachedPolicies(instanceId: string): Promise<IAM.attachedPoliciesListType> {
+    protected async getAttachedPolicies(instanceId: string): Promise<IAM.AttachedPolicy[]> {
         const IamRole = await this.ec2Client.getAttachedIamRole(instanceId)
-        if (!IamRole) {
+        if (!IamRole?.Arn) {
             return []
         }
-        const iamResponse = await this.iamClient.listAttachedRolePolicies(IamRole!.Arn!)
-
-        return iamResponse.AttachedPolicies ?? []
+        try {
+            const attachedPolicies = await this.iamClient.listAttachedRolePolicies(IamRole.Arn)
+            return attachedPolicies
+        } catch (e) {
+            const errorMessage = `No policies attached to role: ${IamRole.Arn}.`
+            getLogger().error(`ec2: ${errorMessage}`)
+            throw ToolkitError.chain(e, errorMessage, { code: 'NoSuchEntity' })
+        }
     }
 
     public async hasProperPolicies(instanceId: string): Promise<boolean> {
-        const attachedPolicies = (await this.getAttachedPolicies(instanceId)).map(policy => policy.PolicyName!)
-        const requiredPolicies = ['AmazonSSMManagedInstanceCore', 'AmazonSSMManagedEC2InstanceDefaultPolicy']
-
-        return requiredPolicies.length !== 0 && requiredPolicies.every(policy => attachedPolicies.includes(policy))
+        try {
+            const attachedPolicies = (await this.getAttachedPolicies(instanceId)).map(policy => policy.PolicyName!)
+            const requiredPolicies = ['AmazonSSMManagedInstanceCore', 'AmazonSSMManagedEC2InstanceDefaultPolicy']
+
+            return requiredPolicies.length !== 0 && requiredPolicies.every(policy => attachedPolicies.includes(policy))
+        } catch (e) {
+            if (e instanceof ToolkitError && e.code == 'NoSuchEntity') {
+                getLogger().warn(
+                    `ec2: due to error in checking policies attached to instance, assuming necessary policies do not exist for instance ${instanceId}.`
+                )
+                return false
+            }
+            throw new ToolkitError(`An unknown error occurred when checking the policies for ${instanceId}`, {
+                cause: e as Error,
```

Comment:
- I think `ToolkitError.chain()` sets `cause`.
- the `code` never needs to have "Error" in its name

    throw new ToolkitError.chain(e, `Failed to check policies for EC2 instance: ${instanceId}`, { code: 'PolicyCheck' })